### PR TITLE
[#1625] More formatter fixes and tests.

### PR DESCRIPTION
**Pull request by ArneDeutsch**
Signed-off-by: Arne Deutsch <Arne.Deutsch@itemis.de>

### CODE REVIEW

**Comment by cdietrich** [comment]
this now does

```
-               //(definesHiddenTokens?='hidden' '(' (hiddenTokens+=[AbstractRule|RuleID] (',' hiddenTokens+=[AbstractRule|RuleID])*)? ')')?
+               //(definesHiddenTokens?='hidden' '(' (hiddenTokens+=[AbstractRule|RuleID] (',' hiddenTokens+=[AbstractRule|RuleID])*)? ')')
+               //?
```

**Comment by cdietrich** [comment]
https://github.com/eclipse/xtext-core/pull/1643

File: org.eclipse.xtext/src/org/eclipse/xtext/xtext/XtextFormatterJava.java [hunk 1]
```diff
@@ -202,13 +209,19 @@ private void formatGroupElements(CompoundElement group, IFormattableDocument doc
...
-			index++;
-			boolean last = index == size;
+		for (int index = 0; index < elements.size(); index++) {
+			boolean last = index + 1 == size;
+			AbstractElement element = elements.get(index);
+			AbstractElement next = null;
+			if (!last)
+				next = elements.get(index + 1);
+			if (first && !last) {
+				if (!(element instanceof UntilToken) && !(element instanceof Group) && !(next instanceof UntilToken))
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/xtext/XtextFormatterJava.java:219]: Can we extract this if condition and give it a name?

**Review by szarnekow — APPROVED** [review]

**Comment by szarnekow** [comment]
see also https://github.com/eclipse/xtext-core/issues/1625#issuecomment-739769256 @ArneDeutsch

File: org.eclipse.xtext/src/org/eclipse/xtext/xtext/XtextFormatterJava.java (hunk 1 above)
Comment (ArneDeutsch) [reply org.eclipse.xtext/src/org/eclipse/xtext/xtext/XtextFormatterJava.java:219]: Extracted a method.

**Pull request merged**